esp-wifi/src/common_adapter/mod.rs

```diff
@@ -308,12 +312,15 @@ pub unsafe extern "C" fn esp_fill_random(dst: *mut u8, len: u32) {
     trace!("esp_fill_random");
     let dst = core::slice::from_raw_parts_mut(dst, len as usize);
 
-    if let Some(ref mut rng) = RANDOM_GENERATOR {
-        for chunk in dst.chunks_mut(4) {
-            let bytes = rng.random().to_le_bytes();
-            chunk.copy_from_slice(&bytes[..chunk.len()]);
+    critical_section::with(|cs| {
```

Comment:
We probably don't want a critical section here, although I'm not sure. Since the RNG is cloneable and we are ignoring its throughput anyway (at least right now), maybe we can just check an `INITED: AtomicBool` flag, and conjure it up ourselves. That, or we should use an RNG-specific lock so that the second core isn't blocked when wifi needs a big bunch of random numbers.

For the single `random()` it probably takes more time to lock than to read the register, which isn't ideal. We could even be impolite and just use RNG without passing it to the stack in the first place.

Similar logic applies to RADIO_CLOCKS if that is ever used between init and deinit.

Reply:
True

RADIO_CLOCKS is used in normal operation if modem-sleep is enabled

Given we require the user to pass them to `initialize` we don't need that `INITED` flag and can just conjure them when needed

Reply:
I like this direction, this opens up the ability to pass `Rng` or `Trng` with a rng trait bound instead of a concrete type.